Fix `await_holding_lock` not linting `parking_lot` Mutex/RwLock

flip1995 opened the pull request:
This adds tests for `RwLock` and `parking_lot::{Mutex, RwLock}`, which were added before in https://github.com/rust-lang/rust-clippy/commit/2dc8c083f54454ca87bb09d691577eada2d23539, but never tested in UI tests. I noticed this while reading [fasterthanli.me](https://fasterthanli.me/articles/a-rust-match-made-in-hell) latest blog post, complaining that Clippy doesn't catch this for `parking_lot`. (Too many people read his blog, he's too powerful)

Some more things:
- Adds a test for #6446
- Improves the lint message

changelog: [`await_holding_lock`]: Now also lints for `parking_lot::{Mutex, RwLock}`

rust-highfive commented:
r? @giraffate

(rust-highfive has picked a reviewer for you, use r? to override)

llogiq commented on tests/ui/await_holding_lock.rs:
```diff
@@ -1,64 +1,194 @@
 #![warn(clippy::await_holding_lock)]
 
-use std::sync::Mutex;
+extern crate parking_lot;
```
Is this needed because we're in a UI test? Elsewhere I'd just `use`.

flip1995 replied:
Not sure. I just copied this from the `let_underscore` test, that also tests for the parking_lot crate. I'll try if it also works with `use`, now that we're using the 2018/2021 (?) edition.

llogiq reviewed (commented):
Not the approver, but this looks good in general. One point: if this still has bugs, we should document them in the lint docs. Otherwise should we re-group this to `suspicious` so it at least warns by default?

flip1995 commented:
> One point: if this still has bugs, we should document them in the lint docs. 

I think the bug for which I added a test is already documented under `Known Problems`, but the documentation could be improved.

> Otherwise should we re-group this to `suspicious` so it at least warns by default?

I'll try to do that and run the `lintcheck` tool.

flip1995 commented:
> I'll try to do that and run the `lintcheck` tool.

I ran the lintcheck tool on the top 279 downloaded crates on crates.io, where grepping for `.await` gave 7645 matches. The `await_holding_lock` lint didn't trigger once. So IMO we could move it to `suspicious`.

Actual changes in this rebase to previous version:
- Moved lints to `suspicious`
- removed `extern crate` from test
- Improved documentation (This is now doc-tested)

llogiq commented:
Great job! Fingers crossed that it'll work as great in production.

@bors r+

bors commented:
:pushpin: Commit ea0ce7bb76bb01741fc3ba2ca32a80c75a380d0f has been approved by `llogiq`

<!-- @bors r=llogiq ea0ce7bb76bb01741fc3ba2ca32a80c75a380d0f -->
<!-- homu: {"type":"Approved","sha":"ea0ce7bb76bb01741fc3ba2ca32a80c75a380d0f","approver":"llogiq"} -->

bors commented:
:hourglass: Testing commit ea0ce7bb76bb01741fc3ba2ca32a80c75a380d0f with merge 02f3c17593e653e30e8835500319ef412855fa34...
<!-- homu: {"type":"BuildStarted","head_sha":"ea0ce7bb76bb01741fc3ba2ca32a80c75a380d0f","merge_sha":"02f3c17593e653e30e8835500319ef412855fa34"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/5246496504?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/5246484523?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/5246682076?check_suite_focus=true)
Approved by: llogiq
Pushing 02f3c17593e653e30e8835500319ef412855fa34 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"llogiq","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/5246682076?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/5246496504?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/5246484523?check_suite_focus=true"},"merge_sha":"02f3c17593e653e30e8835500319ef412855fa34"} -->
